flang/lib/Optimizer/CodeGen/Target.h

```diff
@@ -0,0 +1,108 @@
+//===- Target.h - target specific details -----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Coding style: https://mlir.llvm.org/getting_started/DeveloperGuide/
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef OPTMIZER_CODEGEN_TARGET_H
+#define OPTMIZER_CODEGEN_TARGET_H
+
+#include "mlir/IR/Types.h"
+#include <memory>
+#include <tuple>
+#include <vector>
+
+namespace llvm {
+class Triple;
+} // namespace llvm
+
+namespace fir {
+class KindMapping;
+
+namespace details {
+/// Extra information about how to marshal an argument or return value that
+/// modifies a signature per a particular ABI's calling convention.
+/// Note: llvm::Attribute is not used directly, because its use depends on an
+/// LLVMContext.
+class Attributes {
```

Comment:
Are these attributes meant to be added/used by the Optimizer only or should lowering already create add some of these ?

Reply:
These are intended to be used for lowering to LLVM.

The one caveat is with a FUNCTION return a CHARACTER type. Lowering must distinguish between a plain old argument and the return value being returned through an argument.

That's not to discourage taking broader advantage of argument attributes in lowering in any way. Special alignments, dealing with generalized descriptors, etc. may create wider opportunities.

Reply:
Thanks the explanation, I now see what I see what I need to update to take advantage of this in the bridge.